internal/ingress/controller/store/endpointslice.go

```diff
@@ -34,7 +34,12 @@ func (s *EndpointSliceLister) MatchByKey(key string) ([]*discoveryv1.EndpointSli
 	var eps []*discoveryv1.EndpointSlice
 	// filter endpointSlices owned by svc
 	for _, listKey := range s.ListKeys() {
-		if !strings.HasPrefix(listKey, key) {
+		if len(key) < 58 && !strings.HasPrefix(listKey, key) {
```

Comment:
Where does 58 come from? Can it change elsewhere and resurface this issue in another form?

Reply:
> Where does 58 come from? Can it change elsewhere and resurface this issue in another form?

If the svc is too long, Kubernetes will use the first 58 characters + a random string to form the name of the endpointslices

Reply:
Yeah, I'm concerned about that - what if in next k8s version they change it to say 48 characters?

What is the performance concern here? Is `s.GetByKey(listKey)` costly? How costly? I'd guess it does not even make an API call and just fetches the object from the cache store.

Reply:
@ElvinEfendi slices are using meta generated name "feature" https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/utils.go#L142
this one is full lenght, still not truncated. Generated name is later used for name creation by k8s. As per docs
https://pkg.go.dev/k8s.io/apiserver/pkg/storage/names#NameGenerator.GenerateName
`not exceed the length of a standard Kubernetes name (63 characters)`  . Mentioned 63 chars needs to contain random generated part. Look at https://github.com/kubernetes/apiserver/blob/master/pkg/storage/names/generate.go#L45.
and truncating ifself https://github.com/kubernetes/apiserver/blob/master/pkg/storage/names/generate.go#L51
As you can see, its hardcoded, not optional, at least for now, but I can reference it ( see next commit ). Problem could rise if it becomes optional, but it will be compile error after go k8s deps upgrade.

Regarding optimization, I did test with 10k slices in one namespace and 10k in another. With this optimization, we get rid of second namespace, because keys starts with namespace. Calling `MatchByKey` 10k times, results are 86.85sec and 36.42sec with this optimization. Using real world numbers, 500 slices + 500 slices different ns, 500 calls results in 120.881986ms and optimized 57.799922ms. It make sense to me at least to get rid of different namespaces.

Reply:
Hi. I was the one creating the issue for that. As far as I can see vcluster (which is where this issue appeared for me in the first place) creates the endpointslices itself (as they are synced from the vcluster as far as I know). In that case they are truncated way earlier then 58 characters. As the service contains the name, namespace and cluster name from the vcluster they are long per default (but not longer then 63 chars).

An endpoint slice there looks like this:
```dev-yyyyyy-product-manual-backend-http-x-dev-yyyyyy--f53adt9ggw```

While the matching service looks like this:
```dev-yyyyyy-product-manual-backend-http-x-dev-yyyyyy--f53ad2ba8b```

I know that this might be a special case but it still is possible that someone or something creates endpointslices manually and it would be confusing if they don't work in my opinion.

Reply:
@tombokombo thanks for those numbers, the optimization does make sense in that case.

Also 👍🏼 for referencing `MaxGeneratedNameLength` instead of hardcoding it. It looks good now.